### ISSUE TITLE
Investigate using lightweight isolates for test loading

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/test#1570]
I believe that you have to use `Isolate.spawn` instead of `Isolate.spawnUri` for this to work. That involves a significant body of work but could potentially pay big dividends in terms of test loading speed.

Specifically we would need a single parent isolate that imports all the tests, and drives actually spawning the isolates by calling `Isolate.spawn` on their `main` (or more specifically some function similar to the current bootstrapping).

We man manage a separate send/receive port for each test as well.

Debugging might also get interesting, but should be doable since each test will still have its own isolate.

**Comment by jakemac53** [comment]
Another possible interesting issue we could run into is determining the language version of the parent isolate, and whether it should run with strong null safety enabled or not, etc.

We very well may need multiple parent isolates, one per language version of the tests or something.

**Comment by natebosch** [comment]
We'd need to be careful that any single test file with a syntax error doesn't prevent all tests from running.

**Comment by jakemac53** [comment]
Another benefit of this approach would be only compiling a single app (the parent isolate).

> We'd need to be careful that any single test file with a syntax error doesn't prevent all tests from running.

Hmm ya that is an interesting point, which might be fairly difficult to resolve with this approach. We could ensure the test itself can be parsed (we already parse them anyways) but we can't realistically ensure all its transitive deps are also valid in a short amount of time. Maybe it is enough to just check the test, but there would be situations where you could run some of your tests today and wouldn't be able to in this model. We could fall back on some different behavior if we detect that situation but that would be pretty gross.

Ultimately I think it would likely be worth the tradeoff though, if we could reduce Isolate spawn times by 10x as advertised.

**Comment by natebosch** [comment]
> We could fall back on some different behavior if we detect that situation but that would be pretty gross.

Yeah it might be sufficient to catch an error spawning the isolate and go back to the per-file approach.
